kratos/processes/compute_nodal_gradient_process.cpp

```diff
@@ -32,6 +32,23 @@ void ComputeNodalGradientProcess<THistorical>::Execute()
     // Set to zero
     ClearGradient();
 
+    if (mrModelPart.NumberOfElements() != 0) {
+        ComputeElementalContributionsAndVolume();
+    }
```

Comment:
if you do this check, are the variables properly initialized? => might otherwise screw with the assemble

Reply:
They are set in the constructor, so they should exist

Reply:
Yes, they are initialized to 0 in ClearGradient()

Reply:
But looking at the code:

```c++
KRATOS_ERROR_IF_NOT(mrModelPart.Nodes().begin()->Has(rOriginVariable)) << "Variable " << rOriginVariable.Name() << " not defined on non-historial database" << std::endl;
```

that could be a problem. We should reduce that check so it passes in all ranks if at least one of them has it.

Reply:
Should that be handled inside the macro? Do you have an example of code doing such check?

Reply:
No, the one I am speaking about is the one in the class constructors (~224)

Should look something like:
```c++
if (!mNonHistoricalVariable) {
    // Not sure about this, seems to be doing nothing.
    VariableUtils().CheckVariableExists(rOriginVariable, mrModelPart.Nodes());
} else {
    int hasVariable = 0;

    if(mrModelPart.NumbrOfNodes() != 0)
        hasVariable = mrModelPart.Nodes().begin()->Has(rOriginVariable);
    hasVariable = DataCommunicator().MaxAll(hasVariable);

    KRATOS_ERROR_IF_NOT(hasVariable) << "Variable " << rOriginVariable.Name() << " not defined on non-historial database" << std::endl;
}
```

Reply:
why is it necessary to communicate
IMO is enough to check and throw locally, MPI will then terminate anyway

Reply:
It would be ok to terminate if all of them do the same, but you may want to do the process in a non empty partition, then synchronize globally. If you had a communicator with the non empty partition we could do what you say, but as it is right now in order to communicate A with B you also need to make the call in C, even if it is empty

Reply:
ah sure, I am stupid, sorry for the noise :man_facepalming: